syntax = "proto3";

package storage;

service Storage {
  rpc LoadModelAsync (LoadModelRequest) returns (LoadModelResponse) {}
  rpc ConfirmModel (ConfirmModelRequest) returns (ConfirmModelResponse) {}
  rpc UnloadModel (UnloadModelRequest) returns (UnloadModelResponse) {}

  rpc ClearMem (ClearMemRequest) returns (ClearMemResponse) {}

  rpc RegisterModel (RegisterModelRequest) returns (RegisterModelResponse) {}
  rpc GetServerConfig (GetServerConfigRequest) returns (GetServerConfigResponse) {}

  // Pool memory management
  rpc AllocatePoolMemory (AllocatePoolMemoryRequest) returns (AllocatePoolMemoryResponse) {}
  rpc FreePoolMemory (FreePoolMemoryRequest) returns (FreePoolMemoryResponse) {}
  rpc AllocateFromPool (AllocateFromPoolRequest) returns (AllocateFromPoolResponse) {}
  rpc FreeFromPool (FreeFromPoolRequest) returns (FreeFromPoolResponse) {}
  rpc CopyToPool (CopyToPoolRequest) returns (CopyToPoolResponse) {}
  rpc CopyFromPool (CopyFromPoolRequest) returns (CopyFromPoolResponse) {}
}

message GetServerConfigRequest {
}

message GetServerConfigResponse {
  int64 mem_pool_size = 1;
  int64 chunk_size = 2;
}

message RegisterModelRequest {
  string model_path = 1;
}

message RegisterModelResponse {
  string model_path = 1;
  int64 model_size = 2;
}

enum DeviceType {
  DEVICE_TYPE_DISK = 0;
  DEVICE_TYPE_CPU = 1;
  DEVICE_TYPE_GPU = 2;
}

message MemCopyChunk {
  uint64 src_offset = 1;
  uint64 size = 2;
  uint64 dst_offset = 3;
  uint64 handle_idx = 4;
}
message MemCopyChunkList {
  repeated MemCopyChunk chunks = 1;
}
message MemCopyHandle {
  bytes cuda_ipc_handle = 1;
}
message MemCopyHandleList {
  repeated MemCopyHandle handles = 1;
}

// If the handles has one item, all chunks share the same handle, other each chunk should have one handle
message LoadModelRequest {
  string model_path = 1;
  string replica_uuid = 2;
  map<string, MemCopyChunkList> chunks = 3;
  map<string, MemCopyHandleList> handles = 4;
  DeviceType target_device_type = 5;
}

message LoadModelResponse {
  string model_path = 1;
}

message ConfirmModelRequest {
  string model_path = 1;
  string replica_uuid = 2;
  DeviceType target_device_type = 3;
}

message ConfirmModelResponse {
  string model_path = 1;
  int32 code = 2;
}

message UnloadModelRequest {
  string model_path = 1;
  string replica_uuid = 2;
  DeviceType target_device_type = 3;
}

message UnloadModelResponse {
  string model_path = 1;
  int32 code = 2;
}

message ClearMemRequest {
}

message ClearMemResponse {
}

// Pool Memory Management Messages
message AllocatePoolMemoryRequest {
  string client_id = 1;
  int32 device_id = 2;
  int32 size_mb = 3;
}

message AllocatePoolMemoryResponse {
  string client_id = 1;
  repeated bytes memory_handles = 2;
  int32 size_mb = 3;
  int64 memory_ptr = 4;
  int64 start_offset = 5;
  int64 size_bytes = 6;
  string shared_memory_path = 7;
}

message FreePoolMemoryRequest {
  string client_id = 1;
}

message FreePoolMemoryResponse {
  bool success = 1;
}

message AllocateFromPoolRequest {
  string client_id = 1;
  int32 size_kb = 2;
}

message AllocateFromPoolResponse {
  bool success = 1;
  int32 allocation_id = 2;
  repeated int32 chunk_indices = 3;
  int32 chunk_size = 4;
  string error_message = 5;
}

message FreeFromPoolRequest {
  string client_id = 1;
  int32 allocation_id = 2;
}

message FreeFromPoolResponse {
  bool success = 1;
  string error_message = 2;
}

message CopyToPoolRequest {
  string client_id = 1;
  int32 allocation_id = 2;
  int32 chunk_offset = 3;
  bytes data = 4;
}

message CopyToPoolResponse {
  bool success = 1;
}

message CopyFromPoolRequest {
  string client_id = 1;
  int32 allocation_id = 2;
  int32 chunk_offset = 3;
  int32 size = 4;
}

message CopyFromPoolResponse {
  bool success = 1;
  bytes data = 2;
}